### PR TITLE
WIP: Add complex tag filtering

### LINKED ISSUE #64: [TI-59] summary or'd tag filter
_lash on 2017-01-13T23:02:46Z says:_ 

 When supplying tags for timew summary, only entries containing all of the tags specificed are shown. I propose possibility for or'ing them instead/aswell.

CURRENT:

Entry 1 has keyword A, B
Entry 2 has keyword B, C

timew summary A B matches entry 1
timew summary B A matches entry 1
timew summary B matches both
timew summary A C matches neither
timew summary C A matches neither

PROPOSED:

Entry 1 has keyword A, B
Entry 2 has keyword B, C

timew summary A B matches entry 1
timew summary B A matches entry 1
timew summary B matches both
timew summary A C matches both
timew summary C A matches both

### LINKED ISSUE #209: I want to be able to exclude certain tags from reports
I find the combination of Taskwarrior and Timewarrior really great, but there is a minor inconvenience in the way TImewarrior works. In Taskwarrior--and that doesn't seem to be properly documented--one can use exclude filters with something like `-tag` to filter out tasks that have a particular tag. The same functionality doesn't seem to be supported in Timewarrior, which instead treats `-` in `-tag` as part of the tag name.

Use case: I use my desktop computer mainly for work, but from time to time I do some tasks that are not related to work. Those tasks are a relatively low in number and I tag them `private`. I would like to be able to exclude them from the reports. Currently, the only options are to tag all other tasks as `work` and then use `timew summary work` or to manually subtract the time for private tasks from the totals, which doesn't work for reports. It would be easier if I could just do something like `timew month -private`.

### LINKED ISSUE #358: How to query multiple tags?
timew summary tag1 tag2 searches for intervals with both tag1 and tag2 in the same time, but what if I do several things during the day and just want to exclude a certain tag from the summary or include more than one tag into the summary?

### PR DESCRIPTION
**Pull request by limdingwen**
Continuation of #456. This WIP PR aims to add complex tag filtering, allowing for boolean operators to be done on tags. Example:

Screenshot 2021-11-11 at 12 13 56 AM

TODO:
- [ ] Refactor code such that boolean parser is possible
- [ ] Add boolean parser, based off of taskwarrior complex filtering?
- [x] Add basic OR/- support as a test
- [x] Add feature flag
- [ ] Add documentation
- [ ] Add tests

Would close #209.
Would close #64.
Would close #358.

### CODE REVIEW

**Comment by limdingwen** [comment]
```
src % timew config complexFiltering off     
No changes made.
src % timew summary school OR work :yesterday
No filtered data found in the range 2021-11-10T00:00:00 - 2021-11-11T00:00:00 tagged with OR, school, work.
src % timew config complexFiltering on       
Are you sure you want to change the value of 'complexFiltering' from 'off' to 'on'? (yes/no) yes         
Config file /Users/ignis/.timewarrior/timewarrior.cfg modified.
src % timew summary school OR work :yesterday

Wk  Date       Day Tags      Start      End    Time    Total
W45 2021-11-10 Wed work    0:13:55  0:56:52 0:42:57
                   work   10:37:59 12:04:13 1:26:14
                   school 12:34:33 13:39:19 1:04:46
                   [snip]
```

Feature flag added under `complexFiltering`.

File: src/commands/CmdChart.cpp [hunk 1]
```diff
@@ -100,7 +100,7 @@ int renderChart (
   // Load the data.
   auto filtering = IntervalFilterAndGroup ({
     new IntervalFilterAllInRange ({ filter.start, filter.end }),
-    new IntervalFilterAllWithTags (filter.tags())
+    new IntervalFilterAllWithTags (filter.tags(), rules.getBoolean("complexFiltering"))
```
Comment (lauft) [inline src/commands/CmdChart.cpp:103]: I would not make this a top-level configuration (like e.g. `verbose`), but put it into a `commands` subgroup. I propose `commands.filtering` with values `simple` (= default) and `complex` (or maybe better: `boolean`).

File: src/IntervalFilterAllWithTags.cpp [hunk 2]
```diff
@@ -27,18 +27,59 @@
...
-IntervalFilterAllWithTags::IntervalFilterAllWithTags (std::set <std::string> tags): _tags (std::move(tags))
+IntervalFilterAllWithTags::IntervalFilterAllWithTags (std::set <std::string> tags, const bool complexFiltering): _tags (std::move(tags)), _complexFiltering (complexFiltering)
 {}
 
 bool IntervalFilterAllWithTags::accepts (const Interval& interval)
 {
-  for (auto& tag : _tags)
-  {
-    if (! interval.hasTag (tag))
+  if (_complexFiltering) {
```
Comment (lauft) [inline src/IntervalFilterAllWithTags.cpp:35]: As you have already said, this should not be part of `IntervalFilterAllWithTags`, but rather some function or factory which translates the individual command line arguments into filters or filter groups, like e.g. `IntervalFilterOrGroup`, `IntervalFilterWithoutTagSet`, etc.

File: src/IntervalFilterAllWithTags.cpp [hunk 3]
```diff
@@ -27,18 +27,59 @@
...
+    // TODO: WIP, This does not fully implement boolean logic.
+    // Check if OR exists.
+    bool isOr = false;
+    for (auto& tag : _tags)
     {
-      return false;
+      if (tag == "OR")
```
Comment (lauft) [inline src/IntervalFilterAllWithTags.cpp:41]: Currently, I would opt for lowercase, i.e. `or`

**Review by lauft — COMMENTED** [review]
Nice POC! 👍🏻 See also the comments in the code.

**Comment by lauft** [comment]
Thanks for the update, @limdingwen !

> Based on my understanding, this current version would not support boolean parsing yet due to CLI::getFilter(Range) returning an Interval which does not contain ordered information, and more work needs to be done there. Is it intended such that CLI::getFilter(Range) will return a IntervalFilter instead to aid in this?

This is not just a simple replacement of `CLI::getFilter(...)`, there are also the error cases (e.g. missing id, wrong number of ids) and their output which have to be handled correctly. Moving the parsing into a function of `CLI` is a goal, but only the final step. So at first, I would implement the construction of the filtering within each command, even if this means code duplication at first.

Therefore as a first step, I would split `CLI::getFilter(<default range>) : Interval` at first into `CLI::getRange(<default range>) : Range` and `CLI::getTagSet() : std::set<std::string>`. From there we can then adapt each command for the new filtering.

Depending on the respective command (see table in [this comment](https://github.com/GothenburgBitFactory/timewarrior/pull/456#issuecomment-964854694)) we can then compose the appropriate filtering. For the new filtering another `CLI`-function is required, returning a vector of strings from which we then can create the complex tag filter.

Later all this could be combined into some kind of `IntervalFilterFactory` (and maybe moved to `CLI`), but I would wait with this until we have a full overview how the implementation in each command looks like.

**Comment by sruffell** [comment]
@lauft, @limdingwen : I might have missed it, but I was wondering if the [refactor/filtering](https://github.com/GothenburgBitFactory/timewarrior/tree/refactor/filtering) branch is going to be submitted via a separate PR?

**Comment by lauft** [comment]
@sruffell I am currently doing some performance tests to check whether something deterioated with `refactor/filtering`. If I spot no errors, I will merge it to dev.

**Comment by limdingwen** [comment]
Hey, just an update -- currently doing finals and some other personal projects, would come back to this when I have the time 😄

**Comment by lauft** [comment]
@limdingwen FYI: I have merged the refactoring of interval filtering to `dev`. I'd be happy to see you to pursue this WIP further.

Good luck for your finals! 🤞🏻

**Pull request closed without merging**